lib/src/components/NxTransferList/NxTransferList.tsx

```diff
@@ -38,39 +41,63 @@ export default function NxTransferList<T extends string | number>(props: Props<T
     ...attrs
   } = props;
 
+  if (allowReordering && !Array.isArray(selectedItems)) {
+    throw new TypeError('selectedItems must be an array if allowReordering is true');
+  }
+
+  const selectedItemsArray = Array.isArray(selectedItems) ? selectedItems : Array.from(selectedItems);
+
   const availableItemsCountFormatter = availableItemsCountFormatterProp || defaultAvailableItemsCountFormatter,
       selectedItemsCountFormatter = selectedItemsCountFormatterProp || defaultSelectedItemsCountFormatter;
 
-  const groupedItems = useMemo(() => groupBy(item => selectedItems.has(item.id) ? 'selected' : 'available', allItems),
-          [allItems, selectedItems]),
+  const groupedItems = useMemo(() =>
+        !allowReordering
+          ? groupBy(item => selectedItemsArray.includes(item.id) ? 'selected' : 'available', allItems)
+          : {
+            available: allItems.filter(item => !selectedItemsArray.includes(item.id)),
+            selected: selectedItemsArray
+                .map(item => allItems.find(_item => _item.id === item))
+                .filter(item => typeof item !== 'undefined') as DataItem<T>[]
+          },
+      [allItems, selectedItems]
+      ),
       available = groupedItems.available || [],
       selected = groupedItems.selected || [];
 
-  const availableCount = allItems.length - selectedItems.size,
-      selectedCount = selectedItems.size;
+  const availableCount = allItems.length - selectedItemsArray.length,
+      selectedCount = selectedItemsArray.length;
 
-  function onChange(checked: boolean, id: T) {
-    const newSelectedItems = new Set(selectedItems);
+  const handleOnChangeProp = (array: T[]) => onChangeProp((allowReordering ? array : new Set(array)) as P);
 
-    newSelectedItems[checked ? 'add' : 'delete'](id);
+  function onChange(checked: boolean, id: T) {
+    const newSelectedItemsArray = checked
+      ? [...selectedItemsArray, id]
+      : selectedItemsArray.filter(item => item !== id);
 
-    onChangeProp(newSelectedItems);
+    handleOnChangeProp(newSelectedItemsArray);
   }
 
-  function onSelectAll(idsToAdd: Set<T>) {
-    const newSelectedItems = new Set<T>(selectedItems);
-
-    idsToAdd.forEach(id => { newSelectedItems.add(id); });
+  function onSelectAll(idsToAdd: T[]) {
+    handleOnChangeProp([...selectedItemsArray, ...idsToAdd]);
+  }
 
-    onChangeProp(newSelectedItems);
+  function onUnselectAll(idsToRemove: T[]) {
+    const newSelectedItems = selectedItemsArray.filter(item => !idsToRemove.includes(item));
+    handleOnChangeProp(newSelectedItems);
   }
 
-  function onUnselectAll(idsToRemove: Set<T>) {
-    const newSelectedItems = new Set<T>(selectedItems);
+  function onReorderItem(id: T, direction: 1 | -1) {
+    const itemIndex = selectedItemsArray.findIndex(item => item === id);
```

Comment:
IMO a number of things would be cleaner if `TransferListItem` were aware of its index and communicated that index to `onReorderItem` rather than us having to look it up here

Reply:
> IMO a number of things would be cleaner if `TransferListItem` were aware of its index and communicated that index to `onReorderItem` rather than us having to look it up here

Yeah I agree, I've updated it here: 774e9702b6a1123bac0705354974edd6087775a5